### PR TITLE
feat: elastic search update

### PR DESCRIPTION
**Pull request by PierreThibaudeau**
| Q             | A
| ------------- | ---
| Branch?       | main
| Tickets       | 
| License       | MIT
| Doc PR        | 

Updates elasticsearch from  6.18 to 7.14
deprecates IdentifierExtractorInterface

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -37,7 +37,7 @@
         "doctrine/mongodb-odm": "^2.2",
         "doctrine/mongodb-odm-bundle": "^4.0",
         "doctrine/orm": "^2.6.4 || ^3.0",
-        "elasticsearch/elasticsearch": "^6.0 || ^7.0",
+        "elasticsearch/elasticsearch": "7.14.0",
```
Comment (dunglas) [inline composer.json:40]: Is this mandatory or can we relax the contraint a bit (at least allowing new patch versions)?

File: src/Core/Bridge/Elasticsearch/DataProvider/CollectionDataProvider.php [hunk 2]
```diff
@@ -45,14 +46,25 @@ final class CollectionDataProvider implements ContextAwareCollectionDataProvider
...
+     * @param ResourceMetadataFactoryInterface|ResourceMetadataCollectionFactoryInterface $resourceMetadataFactory
      */
-    public function __construct(Client $client, DocumentMetadataFactoryInterface $documentMetadataFactory, IdentifierExtractorInterface $identifierExtractor, DenormalizerInterface $denormalizer, Pagination $pagination, ResourceMetadataFactoryInterface $resourceMetadataFactory, iterable $collectionExtensions = [])
+    public function __construct(Client $client, DocumentMetadataFactoryInterface $documentMetadataFactory, IdentifierExtractorInterface $identifierExtractor = null, DenormalizerInterface $denormalizer, Pagination $pagination, $resourceMetadataFactory, iterable $collectionExtensions = [])
     {
         $this->client = $client;
         $this->documentMetadataFactory = $documentMetadataFactory;
+
+        if ($this->identifierExtractor) {
+            trigger_deprecation('api-platform', '2.7', sprintf('The service "%s" is deprecated and will be removed in 3.0.', IdentifierExtractorInterface::class));
```
Comment (dunglas) [inline src/Core/Bridge/Elasticsearch/DataProvider/CollectionDataProvider.php:57]:
```suggestion
            trigger_deprecation('api-platform', '2.7', sprintf('Passing an instance of "%s" is deprecated and will not be supported in 3.0.', IdentifierExtractorInterface::class));
```

File: src/Core/Bridge/Elasticsearch/DataProvider/CollectionDataProvider.php [hunk 3]
```diff
@@ -77,10 +89,18 @@ public function supports(string $resourceClass, ?string $operationName = null, a
...
-            $this->identifierExtractor->getIdentifierFromResourceClass($resourceClass);
-        } catch (NonUniqueIdentifierException $e) {
-            return false;
+        if ($this->identifierExtractor) {
+            try {
+                $this->identifierExtractor->getIdentifierFromResourceClass($resourceClass);
+            } catch (NonUniqueIdentifierException $e) {
+                return false;
+            }
+        } else if ($this->resourceMetadataFactory instanceof ResourceMetadataCollectionFactoryInterface) {
```
Comment (dunglas) [inline src/Core/Bridge/Elasticsearch/DataProvider/CollectionDataProvider.php:98]:
```suggestion
        } elseif ($this->resourceMetadataFactory instanceof ResourceMetadataCollectionFactoryInterface) {
```

Comment (dunglas) [reply src/Core/Bridge/Elasticsearch/DataProvider/CollectionDataProvider.php:98]: And what happens if there is no identifiers extractor and no resource metadata factory?

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/Core/Bridge/Elasticsearch/DataProvider/CollectionDataProvider.php (hunk 2 above)
Comment (PierreThibaudeau) [reply src/Core/Bridge/Elasticsearch/DataProvider/CollectionDataProvider.php:57]: @soyuka is going to delete this interface eventually

Comment (dunglas) [reply src/Core/Bridge/Elasticsearch/DataProvider/CollectionDataProvider.php:57]: Ok but the term "service" isn't appropriate. This class can be used without a DIC.

File: src/Core/Bridge/Elasticsearch/DataProvider/CollectionDataProvider.php (hunk 3 above)
Comment (PierreThibaudeau) [reply src/Core/Bridge/Elasticsearch/DataProvider/CollectionDataProvider.php]: You are right, this condition is unnecessary, the check is already done inside the constructor.

File: composer.json [hunk 4]
```diff
@@ -37,7 +37,7 @@
         "doctrine/mongodb-odm": "^2.2",
         "doctrine/mongodb-odm-bundle": "^4.0",
         "doctrine/orm": "^2.6.4 || ^3.0",
-        "elasticsearch/elasticsearch": "^6.0 || ^7.0",
+        "elasticsearch/elasticsearch": "^7.14.0",
```
Comment (dunglas) [inline composer.json:40]:
```suggestion
        "elasticsearch/elasticsearch": "^7.14",
```

**Review by dunglas — APPROVED** [review]

**Comment by dunglas** [comment]
To pass the "Semantic Pull Request" green, replace `feat :` by `feat:` in your first commit message.

**Comment by soyuka** [comment]
thanks @PierreThibaudeau !

**Pull request merged**